File: DailyChallenges/Scala/LeetCode_989.scala
object Solution {
    def addToArrayForm(num: Array[Int], k: Int): List[Int] = {
        def loop(i: Int, k: Int, c: Int, l: List[Int]): List[Int] = 
            if (k>0 || i>=0 || c>0){
                val _c = c + k%10 + (if (i>=0) num(i) else 0)
                loop(i-1, k/10, _c/10, _c%10 :: l)
            }else l
        loop(num.length-1, k, 0, Nil)
    }    
}

File: DailyChallenges/CPP/LeetCode_989.cpp
class Solution {
public:
    vector<int> addToArrayForm(vector<int>& num, int k) {
        vector<int> ans;
        int i = num.size()-1, c = 0;
        while (k > 0 || i>=0 || c>0){
            c += k%10 + (i>=0 ? num[i--] : 0);
            ans.push_back(c%10);
            c /= 10; k /= 10;
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};
